#!/bin/bash
#$ -V
#$ -cwd
#$ -o "logs/Metagenomics_16s_picrust_log.txt"
#$ -j y
#$ -l mem_free=15G,h_vmem=25G
#$ -pe threaded 12



echo "--------------------------------------------------"
echo "Script starts"
date
echo "Based on tutorial posted here:"
echo "http://picrust.github.io/picrust/tutorials/metagenome_prediction.html#metagenome-prediction-tutorial"
echo "--------------------------------------------------"

echo "--------------------------------------------------"
echo "[01]. Loading variables and QIIME env.----------"
echo "--------------------------------------------------"
date
source config.txt
module load qiime/1.9.1

date


echo "--------------------------------------------------"
echo "[02]. Closed refere OTU picking-------------------"
echo "------with 13_5 GreenGenes database---------------"
echo "--------------------------------------------------"
# Closed reference OTU picking, 97% similarity, only GreenGenes possible for PICRUSt predictions.
date
pick_closed_reference_otus.py -i ${SEQS_CLEAR} -o ${PICRUST_DIR}/closed_ref_otus/ -p ${QIIME_PARAMS_PICRUST} -r ${REF_FILE_PIC} -t ${TAXONOMY_FILE_PIC} -a -O 12
pick_closed_reference_otus.py -i ${INPUT_FASTA} -o ${PICRUST_DIR}/closed_ref_otus_usearch/ -p ${QIIME_PARAMS} -r ${REF_FILE} -t ${TAXONOMY_FILE} -a -O 12

date

echo "--------------------------------------------------"
echo "[03]. Loading PICRUSt environment. ---------------"
echo "--------------------------------------------------"
date
module load picrust
date

echo "--------------------------------------------------"
echo "[04]. Normalizing OTU Table-----------------------"
echo "--------------------------------------------------"
date
# First convert uncross produced otu table to biom (to json format)
time normalize_by_copy_number.py -i ${PICRUST_DIR}closed_ref_otus/otu_table.biom -o ${PICRUST_DIR}closed_ref_otus/otu_table_normalized.biom
date

echo "--------------------------------------------------"
echo "[05]. Predict Metagenome Composition--------------"
echo "-based on ancestral states algorithm--------------"
echo "--------------------------------------------------"
date
time predict_metagenomes.py -i ${PICRUST_DIR}closed_ref_otus/otu_table_normalized.biom -o ${PICRUST_DIR}closed_ref_otus/otu_table_normalized_predicted.biom -a ${PICRUST_DIR}closed_ref_otus/otu_table_normalized_predicted_NSTIvals.tab
date

echo "--------------------------------------------------"
echo "[06]. Collapse predictions into KEGG pathways-----"
echo "--------------------------------------------------"
date
categorize_by_function.py -i ${PICRUST_DIR}closed_ref_otus/otu_table_normalized_predicted.biom -c "KEGG_Pathways" -l 3 -o ${PICRUST_ANALYSIS}metagenome_at_level3.biom
categorize_by_function.py -i ${PICRUST_DIR}closed_ref_otus/otu_table_normalized_predicted.biom -c KEGG_Pathways -l 2 -o ${PICRUST_ANALYSIS}metagenome_at_level2.biom
date

echo "--------------------------------------------------"
echo "[07]. Summarizing KEGG pathways through plots-----"
echo "--------------------------------------------------"
date
#through qiime functions
summarize_taxa_through_plots.py -i ${PICRUST_ANALYSIS}metagenome_at_level3.biom -p ${QIIME_PARAMS_PICRUST_l3} -o ${PICRUST_ANALYSIS}plots_lvl3 -c ${MAPPING_METADATA} -m ${MAPPING}
summarize_taxa_through_plots.py -i ${PICRUST_ANALYSIS}metagenome_at_level2.biom -p ${QIIME_PARAMS_PICRUST_l2} -o ${PICRUST_ANALYSIS}plots_lvl2 -c ${MAPPING_METADATA} -m ${MAPPING}
date

echo "--------------------------------------------------"
echo "[08]. Partition metagenome functional contributions"
echo "-according to function, OTU, sample - for a given OTU table"
echo "--------------------------------------------------"
date
echo "      Output is a tab-delimited column indicating OTU contribution to each function."
echo "      Input is an normalized OTU table, BEFORE PICRUSt preditions!"
metagenome_contributions.py -i ${PICRUST_DIR}closed_ref_otus/otu_table_normalized.biom -o ${PICRUST_ANALYSIS}/metagenome_contributions.tab --type_of_prediction ko
#for lupus predictions - K11089
metagenome_contributions.py -i ${PICRUST_DIR}closed_ref_otus/otu_table_normalized.biom -o ${PICRUST_ANALYSIS}/metagenome_contributions.tab --type_of_prediction ko -l K11089
echo "Read more at: https://github.com/mlangill/microbiome_helper/wiki/PICRUSt-tutorial"
date

echo "--------------------------------------------------"
echo "[09]. Group significance comparisons--------------"
echo "--------------------------------------------------"
date
# inputs are "otu" tables, generated by summarize_taxa_through plots
group_significance.py -i ${PICRUST_ANALYSIS}plots_lvl2/metagenome_at_level2_L2.txt -m ${MAPPING} -c ${MAPPING_METADATA} -o ${PICRUST_ANALYSIS}/group_significance_l2_gTest.txt --metadata_key "KEGG_Pathways" -s g_test
group_significance.py -i ${PICRUST_ANALYSIS}plots_lvl3/metagenome_at_level3_L3.txt -m ${MAPPING} -c ${MAPPING_METADATA} -o ${PICRUST_ANALYSIS}/group_significance_l3_gTest.txt --metadata_key "KEGG_Pathways" -s g_test

group_significance.py -i ${PICRUST_ANALYSIS}plots_lvl2/metagenome_at_level2_L2.txt -m ${MAPPING} -c ${MAPPING_METADATA} -o ${PICRUST_ANALYSIS}/group_significance_l2_kruskal_wallis.txt --metadata_key "KEGG_Pathways" -s kruskal_wallis
group_significance.py -i ${PICRUST_ANALYSIS}plots_lvl3/metagenome_at_level3_L3.txt -m ${MAPPING} -c ${MAPPING_METADATA} -o ${PICRUST_ANALYSIS}/group_significance_l3_kruskal_wallis.txt --metadata_key "KEGG_Pathways" -s kruskal_wallis

date

echo "--------------------------------------------------"
echo "[10]. Further analysis----------------------------"
echo "--------------------------------------------------"
date
echo "Further analysis is made using STAMP free software."
echo "For the simplicity, and reproducibility a virtual machine \"Microbiome Helper\" is used, see link:"
echo "https://github.com/mlangill/microbiome_helper"
echo "or paper:"
echo "Comeau AM, Douglas GM, Langille MGI. 2017. Microbiome Helper: a Custom and Streamlined Workflow for Microbiome Research. mSystems 2(1): e00127-16; DOI: 10.1128/mSystems.00127-16"
echo "The inputs are"
biom convert -i ${PICRUST_DIR}closed_ref_otus/otu_table_normalized_predicted.biom -o ${PICRUST_DIR}closed_ref_otus/otu_table_normalized_predicted.txt --to-tsv #--header-key KEGG_Pathways  # (or COG_Category)
echo "load ${PICRUST_DIR}closed_ref_otus/otu_table_normalized_predicted.txt to analyze statistically significant KEGG Orthologs (KO) between groups in mapping file"
echo "Remember to delete first line - \"Constructed from OTU table \" as it would produce errors in PICRUSt"

# for profile preparation you CANNOT have header keys, as it would produce errors in STAMP, but for other analysis it might be handy... just might
echo "Remember to delete first line - \"Constructed from OTU table \" as it would produce errors in PICRUSt"

STAMP_INPUT="${PICRUST_DIR}closed_ref_otus/otu_table_normalized_predicted.txt"
echo "1) predicted metagenome biom table in txt format:"
echo "${STAMP_INPUT}"
echo "2) Mapping file"
echo "${MAPPING}"

echo "See more help at: http://picrust.github.io/picrust/tutorials/stamp_tutorial.html#stamp-tutorial"


echo "--------------------------------------------------"
echo "Script ends---------------------------------------"
echo "--------------------------------------------------"
date
